### PR TITLE
PYTHON-3841 Add durations to connection pool events

### PR DESCRIPTION
**Pull request by NoahStapp**

### CODE REVIEW

**Review by caseyclements — COMMENTED** [review]
Some comments

File: pymongo/monitoring.py [hunk 1]
```diff
@@ -1090,8 +1108,8 @@ class ConnectionCheckOutFailedEvent(_ConnectionEvent):
 
     __slots__ = ("__reason",)
 
-    def __init__(self, address: _Address, reason: str) -> None:
-        super().__init__(address)
+    def __init__(self, address: _Address, reason: str, duration: float) -> None:
+        super().__init__(address, 0, duration)
```
Comment (caseyclements) [inline pymongo/monitoring.py:1112]: Using positional arguments here might not be the best idea. Not only can it be hard to follow, but issues with method resolution order (MRO) but creep in.

File: test/test_connection_monitoring.py [hunk 2]
```diff
@@ -426,13 +426,15 @@ def assertRepr(self, obj):
     def test_events_repr(self):
         host = ("localhost", 27017)
         self.assertRepr(ConnectionCheckedInEvent(host, 1))
```
Comment (caseyclements) [inline test/test_connection_monitoring.py:428]: Similar question from my last review. What is assertRepr? Google didn't even return results for it.

**Review by caseyclements — COMMENTED** [review]
Comments

File: pymongo/monitoring.py (hunk 1 above)
Comment (NoahStapp) [reply pymongo/monitoring.py:1112]: You're suggesting modifying the constructor to keyword args?

File: pymongo/logger.py [hunk 3]
```diff
@@ -128,7 +128,7 @@ def _redact(self) -> None:
         if self._redacted:
             return
         self._kwargs = {k: v for k, v in self._kwargs.items() if v is not None}
-        if "durationMS" in self._kwargs:
+        if "durationMS" in self._kwargs and hasattr(self._kwargs["durationMS"], "total_seconds"):
             self._kwargs["durationMS"] = self._kwargs["durationMS"].total_seconds() * 1000
```
[the remark this answers is not in the document]
Comment (NoahStapp) [reply pymongo/logger.py:132]: `durationMS` is also used by the command logging messages, which use `timedelta`. I believe we can update these instances to use monotonic times as well.

File: test/test_connection_monitoring.py (hunk 2 above)
Comment (caseyclements) [reply test/test_connection_monitoring.py:428]: That's so unusual. What does repr(obj) look like?

File: pymongo/monitoring.py (hunk 1 above)
Comment (caseyclements) [reply pymongo/monitoring.py]: I'm suggesting calling super using keywords. Then you can mess up the order. My intuition is that this will solve potential MRO issues and if not, at least make clear your intentions if someone needs to fix them in the future.
```
class Baz(Bar):
    def __init__(self, foo, bar, baz):
        super().__init__(bar=bar, foo=foo)
        self.baz = baz
```

**Review by caseyclements — APPROVED** [review]
Checks out.

File: test/test_connection_monitoring.py (hunk 2 above)
Comment (ShaneHarvey) [reply test/test_connection_monitoring.py:428]: This is pretty standard to test that repr works as expected according to the docs:
> Return a string containing a printable representation of an object. For many types, this function makes an attempt to return a string that would yield an object with the same value when passed to [eval()](https://docs.python.org/3/library/functions.html#eval);

https://docs.python.org/3/library/functions.html#repr

File: pymongo/logger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply pymongo/logger.py:132]: I'm wondering if we can put a TODO or open a ticket to make this more type safe.

File: pymongo/monitoring.py [hunk 4]
```diff
@@ -1002,6 +1002,24 @@ def __repr__(self) -> str:
...
+
+    __slots__ = ("__duration",)
+
+    def __init__(self, address: _Address, connection_id: int, duration: Optional[float]) -> None:
+        super().__init__(address, connection_id)
+        self.__duration = duration
+
+    @property
+    def duration(self) -> Optional[float]:
+        """The duration of the connection event."""
```
Comment (ShaneHarvey) [inline pymongo/monitoring.py:1016]: Need to add `.. versionadded:: 4.7`

**Review by ShaneHarvey — COMMENTED** [review]
Don't forget to update the changelog with links to the new duration property for each class.

**Review by ShaneHarvey — APPROVED** [review]

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Just noticed there's one failing test: 
```
 [2024/03/04 09:47:14.537] self = <test.test_connection_logging.TestConnectionLoggingConnectionLogging testMethod=test_Connection_checkout_fails_due_to_error_establishing_connection>
 [2024/03/04 09:47:14.537]     def test_case(self):
 [2024/03/04 09:47:14.537] >       self.run_scenario(spec)
 [2024/03/04 09:47:14.537] test/unified_format.py:1854: 
 [2024/03/04 09:47:14.537] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 [2024/03/04 09:47:14.537] test/unified_format.py:1801: in run_scenario
 [2024/03/04 09:47:14.537]     self._run_scenario(spec, uri)
 [2024/03/04 09:47:14.537] test/unified_format.py:1828: in _run_scenario
 [2024/03/04 09:47:14.537]     self.check_log_messages(spec["operations"], expect_log_messages)
 [2024/03/04 09:47:14.537] test/unified_format.py:1762: in check_log_messages
 [2024/03/04 09:47:14.537]     self.match_evaluator.match_result(expected_data, actual_data)
 [2024/03/04 09:47:14.537] test/unified_format.py:791: in match_result
 [2024/03/04 09:47:14.537]     return self._match_document(expectation, actual, is_root=not in_recursive_call)
 [2024/03/04 09:47:14.537] test/unified_format.py:780: in _match_document
 [2024/03/04 09:47:14.537]     self.match_result(value, actual[key], in_recursive_call=True)
 [2024/03/04 09:47:14.537] test/unified_format.py:808: in match_result
 [2024/03/04 09:47:14.537]     self.test.assertEqual(expectation, actual)
 [2024/03/04 09:47:14.537] E   AssertionError: 'Connection pool cleared' != 'Connection closed'
 [2024/03/04 09:47:14.537] E   - Connection pool cleared
 [2024/03/04 09:47:14.537] E   ?            -----  ^^^
 [2024/03/04 09:47:14.537] E   + Connection closed
 [2024/03/04 09:47:14.537] E   ?              ^^

```
https://spruce.mongodb.com/task/mongo_python_driver_test_linux_enterprise_auth__platform~rhel8_auth~auth_python_version~3.10_test_enterprise_auth_patch_dc50cbd35ad74200d1f8eb9bbba1e3a624b53514_65e6044032f417dd385ee9a9_24_03_04_17_26_26?execution=0&sortBy=STATUS&sortDir=ASC

**Comment by NoahStapp** [comment]
> Just noticed there's one failing test:
> 
> ```
>  [2024/03/04 09:47:14.537] self = <test.test_connection_logging.TestConnectionLoggingConnectionLogging testMethod=test_Connection_checkout_fails_due_to_error_establishing_connection>
>  [2024/03/04 09:47:14.537]     def test_case(self):
>  [2024/03/04 09:47:14.537] >       self.run_scenario(spec)
>  [2024/03/04 09:47:14.537] test/unified_format.py:1854: 
>  [2024/03/04 09:47:14.537] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
>  [2024/03/04 09:47:14.537] test/unified_format.py:1801: in run_scenario
>  [2024/03/04 09:47:14.537]     self._run_scenario(spec, uri)
>  [2024/03/04 09:47:14.537] test/unified_format.py:1828: in _run_scenario
>  [2024/03/04 09:47:14.537]     self.check_log_messages(spec["operations"], expect_log_messages)
>  [2024/03/04 09:47:14.537] test/unified_format.py:1762: in check_log_messages
>  [2024/03/04 09:47:14.537]     self.match_evaluator.match_result(expected_data, actual_data)
>  [2024/03/04 09:47:14.537] test/unified_format.py:791: in match_result
>  [2024/03/04 09:47:14.537]     return self._match_document(expectation, actual, is_root=not in_recursive_call)
>  [2024/03/04 09:47:14.537] test/unified_format.py:780: in _match_document
>  [2024/03/04 09:47:14.537]     self.match_result(value, actual[key], in_recursive_call=True)
>  [2024/03/04 09:47:14.537] test/unified_format.py:808: in match_result
>  [2024/03/04 09:47:14.537]     self.test.assertEqual(expectation, actual)
>  [2024/03/04 09:47:14.537] E   AssertionError: 'Connection pool cleared' != 'Connection closed'
>  [2024/03/04 09:47:14.537] E   - Connection pool cleared
>  [2024/03/04 09:47:14.537] E   ?            -----  ^^^
>  [2024/03/04 09:47:14.537] E   + Connection closed
>  [2024/03/04 09:47:14.537] E   ?              ^^
> ```
> 
> [spruce.mongodb.com/task/mongo_python_driver_test_linux_enterprise_auth__platform~rhel8_auth~auth_python_version~3.10_test_enterprise_auth_patch_dc50cbd35ad74200d1f8eb9bbba1e3a624b53514_65e6044032f417dd385ee9a9_24_03_04_17_26_26?execution=0&sortBy=STATUS&sortDir=ASC](https://spruce.mongodb.com/task/mongo_python_driver_test_linux_enterprise_auth__platform~rhel8_auth~auth_python_version~3.10_test_enterprise_auth_patch_dc50cbd35ad74200d1f8eb9bbba1e3a624b53514_65e6044032f417dd385ee9a9_24_03_04_17_26_26?execution=0&sortBy=STATUS&sortDir=ASC)

Fixed, forgot to re-order the spec test ordering for the pool close vs pool clear events.

**Review by ShaneHarvey — APPROVED** [review]
